### PR TITLE
[sc-5170] add retry mechanism

### PR DESCRIPTION
**Pull request by steb-veksler**

### CODE REVIEW

File: cmd/deploy.go [hunk 1]
```diff
@@ -290,11 +294,19 @@ func installHelmRelease(ctx context.Context, helmClient *helm.Client, releaseNam
 	spinner := ui.NewSpinner("Installing groundcover helm release")
 	spinner.Start()
 	spinner.StopMessage("groundcover helm release is installed")
+	spinner.StopFailMessage("groundcover helm release installation failed")
 	defer spinner.Stop()
 
 	var release *helm.Release
-	if release, err = helmClient.Upgrade(ctx, releaseName, chart, chartValues); err != nil {
-		spinner.StopFailMessage("groundcover helm release installation failed")
+	helmUpgradeFunc := func() error {
```
Comment (maxlevinps) [inline cmd/deploy.go:301]: define the funcs outside

File: cmd/status.go [hunk 2]
```diff
@@ -120,7 +121,7 @@ func waitForPortal(ctx context.Context, kubeClient *k8s.Client, helmRelease *hel
 
 	appVersion := helmRelease.Chart.AppVersion()
 
-	isPortalRunningFunc := func() (bool, error) {
+	isPortalRunningFunc := func() error {
```
Comment (maxlevinps) [inline cmd/status.go:124]: same here,  define it outside this func

File: cmd/status.go [hunk 3]
```diff
@@ -165,18 +167,24 @@ func waitForAlligators(ctx context.Context, kubeClient *k8s.Client, helmRelease
 
 	runningAlligators := 0
 
-	isAlligatorRunningFunc := func() (bool, error) {
+	isAlligatorRunningFunc := func() error {
```
Comment (maxlevinps) [inline cmd/status.go:170]: same here,  define it outside this func

File: pkg/ui/spinner.go [hunk 4]
```diff
@@ -60,13 +71,22 @@ func (s *Spinner) Poll(ctx context.Context, function func() (bool, error), inter
...
 		case <-ticker.C:
-			success, err := function()
-			if err != nil {
+			err := function()
+
+			if err == nil {
+				return nil
+			}
+
+			var rerr *retryableError
```
Comment (maxlevinps) [inline pkg/ui/spinner.go:80]: use `retryableErr`

File: pkg/ui/spinner_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+func (suite *SpinnerTestSuite) SetupSuite() {}
+
+func (suite *SpinnerTestSuite) TearDownSuite() {}
+
+func TestSpinnerSuite(t *testing.T) {
+	suite.Run(t, &SpinnerTestSuite{})
+}
+
+func (suite *SpinnerTestSuite) TestPollFuncSuccues() {
```
Comment (maxlevinps) [inline pkg/ui/spinner_test.go:25]: see if you can use `https://github.com/benbjohnson/clock` lib instead of really waiting

**Review by maxlevinps — CHANGES_REQUESTED** [review]

File: cmd/deploy.go (hunk 1 above)
Comment (steb-veksler) [reply cmd/deploy.go:301]: i need the func on same scope to set release variable, spinner.Poll except func() error

File: cmd/status.go (hunk 3 above)
Comment (steb-veksler) [reply cmd/status.go:170]: i need the func on same scope to set runningAlligator variable, spinner.Poll except func() error

File: pkg/ui/spinner_test.go (hunk 5 above)
Comment (steb-veksler) [reply pkg/ui/spinner_test.go:25]: https://github.com/benbjohnson/clock#controlling-time
you need to use mock.Ticker which will not test the function

**Review by maxlevinps — APPROVED** [review]

**Pull request merged**